documentation/correctness.adoc

```diff
@@ -0,0 +1,123 @@
+**Note** that this is a condition that pertains to the deserializing client, rather than on the serialization chunk.
+Incorrectness could be partially recovered from by partially deserializing into "`dynamic`"/"`reflective`" nodes.
+
+=== Meta-structural
+
+The seralization matches the metamodels - in particular:
+
+* Concept identifiers (tuples with a `key` value) are uniquely resolvable to a concept in a metamodel.
+Incorrectness is to be reported in terms of the ID of the serialized and the key of the concept which can not be resolved.
+* For every object corresponding to a serialized node:
+** For every entry of the `properties`, `children`, or `references` arrays, the `key` in the `property`, `containment`, resp. `reference` sub object is the key of a feature of the concept pointed to by the `concept` identifier.
+Incorrectness is to be reported in terms of the ID of the serialized node, the feature type (property/containment/reference), the index of the offending entry, and the key which can not be resolved.
+** For every entry of a `properties` array, the value of the `value` property conforms to the `type` specified by the `Property` feature corresponding to the key.
+
+Incorrectness is to be reported in terms of the ID of the serialized node, the feature type (property/containment/reference), the index of the offending entry, and the key which can not be resolved.
+"`To resolve a key`" means that we can find a concept or feature in some metamodel with that key, and with (exactly-)matching version.
+Incorrectness is to be reported as a model with references to nodes, as well as an appropriate error message.
+Incorrectness could be partially recovered from by partially deserializing into "`dynamic`"/"`reflective`" nodes, leaving the unresolvable meta-references as-is.
+
+
+== Model (post-deserialization/in-memory)
+
+=== Structure-referential
+
+All links (on M1/instance-level) resolve.
+Note that this doesn't require any knowledge of the metamodels, as it just involves nodes' IDs.
+This means that the serialization chunk doesn't need to be correct in the meta-referential nor in the meta-structural sense!
+It should be alway possible to deserialize the serialization chunk into "`dynamic`"/"`reflective`" nodes, which could be useful in its own right.
+
+**TODO:** describe when and how links should be resolved, etc.
+
+=== Constraints
+
+No constraint violations.
+Constraints can only be checked with full knowledge of all metamodels involved, so meta-referential, meta-structural correctness is necessary for this.
+Also: the more "`interesting`" constraints typically navigate across references, so structure-referential correctness would also be necessary.
+It should be possible to come up with some kind of "`partial constraint checking`" that deals with unresolved references in a meaningful way -- possibly reporting warnings, rather than outright constraint violations.
```

Comment:
Is this connected to the partial evaluation talk :D ?
This may be a bit too advanced for me to implement

Reply:
Yes, it is :) Partial evaluation is actually not that hard to implement - as my talk will show ;)

Reply:
Maybe we should remember Federico's mantra: "LionWeb does not want to be innovative, but the boring base to innovate upon."

Especially validations should also be easy to read/understand, as users might not understand why some error is flagged and then try to read/debug the validation implementation. (Some old nightmares from EMF validation / Xtext scoping are resurfacing ...)

Reply:
That's true, which is also the reason I need some more time for that proposal.